File: packages/api/src/management/entitlements.ts
import { z } from "zod";

const featureSchema = z.object({
  multiOrganization: z.boolean(),
  cloudLanding: z.boolean(),
  cloudInfrastructure: z.boolean(),
  openSignUp: z.boolean(),
  clickhouseAudit: z.boolean(),
  auditEnterpriseStreams: z.boolean(),
  complianceExport: z.boolean(),
});

const limitSchema = z.object({
  organizations: z.number().nullable(),
  nodes: z.number().nullable(),
  seats: z.number().nullable(),
  relays: z.number().nullable(),
});

export const entitlementsSchema = z.object({
  status: z.enum(["community", "active", "grace", "expired"]),
  tier: z.enum(["community", "cloud", "enterprise"]),
  features: featureSchema,
  limits: limitSchema,
  licenseId: z.string().nullable(),
  subject: z.string().nullable(),
  issuedAt: z.number().nullable(),
  notAfter: z.number().nullable(),
  graceUntil: z.number().nullable(),
  stale: z.boolean(),
  reason: z.string().nullable(),
});

export type EntitlementsResponse = z.infer<typeof entitlementsSchema>;

File: crates/tunnet-license/src/entitlements.rs
use crate::error::LicenseFailureCode;
use crate::features::{Entitlements, Feature, LicenseStatus, community_with_reason};
use crate::verify::{License, RuntimeStatus};

pub use crate::features::community;

pub fn entitlements_from(license: &License, status: RuntimeStatus, stale: bool) -> Entitlements {
    let status = match status {
        RuntimeStatus::Active => LicenseStatus::Active,
        RuntimeStatus::Grace => LicenseStatus::Grace,
    };
    Entitlements {
        status,
        tier: license.tier,
        features: license.features,
        limits: license.limits,
        license_id: Some(license.jti.clone()),
        subject: Some(license.sub.clone()),
        issued_at: Some(license.iat),
        not_after: Some(license.exp),
        grace_until: if license.grace > 0 {
            Some(license.exp + license.grace)
        } else {
            None
        },
        stale,
        reason: None,
    }
}

pub fn has_feature(entitlements: &Entitlements, feature: Feature) -> bool {
    entitlements.features.get(feature)
}

pub fn fallback_community(reason: LicenseFailureCode, stale: bool) -> Entitlements {
    community_with_reason(Some(reason), stale)
}
